include/dali/core/tensor_shape.h

```diff
@@ -1086,51 +1086,219 @@ auto collapse_dim(const TensorShape<ndim> &shape, int dim_idx) {
  *
- * If the dimension is not covered by the shape_blocks it is not collapsed.
+ * If the dimension is not covered by the dim_groups it is not collapsed.
  *
  * Allows to collapse more the one dimension in one go.
- * For example shape = [2, 4, 10, 30, 3]; shape_blocks = {{0, 2}, {2, 3}} will return: [8, 90].
+ * For example shape = [2, 4, 10, 30, 3]; dim_groups = {{0, 2}, {2, 3}} will return: [8, 90].
  *
- * @param shape Shape to be collapsed
- * @param shape_blocks Description of blocks in shape {starting_dimension_idx, length}.
- *                     Expected to be sorted, does not require entries describing blocks of
- *                     length 1.
+ * @param shape      Shape to be collapsed
+ * @param dim_groups Description of dimension groups in `shape` {starting_dimension_idx, size}.
+ *                   Unlisted dimensions are implictly preserved (as if they were in a group
+ *                   of size 1)
  * @return Collapsed shape
  */
 template <int out_ndim = DynamicDimensions, int ndim>
 TensorShape<out_ndim> collapse_dims(const TensorShape<ndim> &shape,
-                                    span<const std::pair<int, int>> shape_blocks) {
+                                    span<const std::pair<int, int>> dim_groups) {
   TensorShape<out_ndim> result;
-  int covered_elements = 0;
-  for (const auto &block : shape_blocks) {
-    covered_elements += block.second;
-  }
-  int result_length = shape_blocks.size() + shape.size() - covered_elements;
-  result.resize(result_length);
-  int current_src_dim = 0;
-  int block_idx = 0;
-  for (int i = 0; i < result_length; i++) {
-    if (block_idx < shape_blocks.size() && current_src_dim == shape_blocks[block_idx].first) {
-      assert(current_src_dim == shape_blocks[block_idx].first);
-      result[i] = 1;
-      auto block_start = shape_blocks[block_idx].first;
-      auto block_end = shape_blocks[block_idx].first + shape_blocks[block_idx].second;
-      for (int j = block_start; j < block_end; j++) {
-        result[i] *= shape[j];
+  const int in_dim = shape.sample_dim();
+  int collapsed = 0;
+  for (const auto &group : dim_groups) {
+    assert(group.first >= 0 && group.first < in_dim && "dimension index out of range");
+    assert(group.second > 0 && group.first + group.second <= in_dim && "invalid group size");
+    collapsed += group.second - 1;
+  }
+  const int out_dim = in_dim - collapsed;
+  assert(out_ndim < 0 || out_ndim == out_dim);
+  result.resize(out_dim);
+  int in_d = 0;
+  int out_d = 0;
+  for (auto group : dim_groups) {
+    assert(group.first >= in_d && "dim_groups not sorted");
+    while (in_d < group.first) {
+      result[out_d++] = shape[in_d++];
+    }
+    int64_t v = shape[in_d++];
+    while (in_d < group.first + group.second) {
+      v *= shape[in_d++];
+    }
+    result[out_d++] = v;
+  }
+  while (in_d < in_dim) {
+    result[out_d++] = shape[in_d++];
+  }
+  assert(out_d == out_dim);
+  assert(in_d == in_dim);
+  return result;
+}
```

Comment:
Looking ad the collapse_dims for TensorListShape, maybe it would be better to wrap the main part as `collapse_dims(span out, span in, groups)` and just call it once here and in a loop for TensorListShape?